README.md
# color_parser

A package to parse the color.
This package is inspired from the JS library https://www.w3schools.com/lib/w3color.js

# Usage

Import `import 'package:color_parser/color_parser.dart';`

Example:

```dart
import 'package:color_parser/color_parser.dart';

ColorParser parser;
// there are various ways to instantiate the object

// using RGB
parser = ColorParser.rgb(0, 191, 255);

// using HexCode
parser = ColorParser.hex('#00bfff');

// using ARGB, A = alpha
parser = ColorParser.argb(255, 0, 191, 255);

// using RGBO, O = opacity
parser = ColorParser.rgbo(0, 191, 255, 1);

// using color value
parser = ColorParser.value(0xff00bff);

// using color itself
parser = ColorParser.color(Colors.lightBlueAccent);

// get the info from the object such as

// color name
print(parser.toName()); // e.g. DeepSkyBlue

// material color
Color color = parser.getColor();

// RGB
print(parser.toRGBString()); // e.g. rgb(0, 191, 255)

// Hex
print(parser.toHex()); // e.g. #00bfff

// HSL
print(parser.toHSLString()); // e.g hsl(195, 100%, 50%)

// HWB
print(parser.toHWBString()); // e.g. hwb(195, 0%, 0%)

// CMYK
print(parser.toCMYKString()); // e.g. cmyk(100%, 25%, 0%, 0%)

// Ncol
print(parser.toNcolString()); // e.g. C25, 0%, 0%
```

There are methods available for String representation as well as for values to use in the code.

# Working Demo:
check the working demo at:
https://deepak786.github.io/demo-color-parser/


CHANGELOG.md
## [0.0.3] - 2021-04-18
* migrated to null safety

## [0.0.2] - 2020-09-11
* Updated description of package

## [0.0.1] - 2020-09-11
* First Release


test/color_parser_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:color_parser/color_parser.dart';

void main() {
}


lib/color_names.dart
Map<String, String> colorNames = {
  'f0f8ff': "AliceBlue",
  'faebd7': "AntiqueWhite",
  '00ffff': "Aqua/Cyan",
  '7fffd4': "Aquamarine",
  'f0ffff': "Azure",
  'f5f5dc': "Beige",
  'ffe4c4': "Bisque",
  '000000': "Black",
  'ffebcd': "BlanchedAlmond",
  '0000ff': "Blue",
  '8a2be2': "BlueViolet",
  'a52a2a': "Brown",
  'deb887': "BurlyWood",
  '5f9ea0': "CadetBlue",
  '7fff00': "Chartreuse",
  'd2691e': "Chocolate",
  'ff7f50': "Coral",
  '6495ed': "CornflowerBlue",
  'fff8dc': "Cornsilk",
  'dc143c': "Crimson",
  '00008b': "DarkBlue",
  '008b8b': "DarkCyan",
  'b8860b': "DarkGoldenRod",
  'a9a9a9': "DarkGray/DarkGrey",
  '006400': "DarkGreen",
  'bdb76b': "DarkKhaki",
  '8b008b': "DarkMagenta",
  '556b2f': "DarkOliveGreen",
  'ff8c00': "DarkOrange",
  '9932cc': "DarkOrchid",
  '8b0000': "DarkRed",
  'e9967a': "DarkSalmon",
  '8fbc8f': "DarkSeaGreen",
  '483d8b': "DarkSlateBlue",
  '2f4f4f': "DarkSlateGray/DarkSlateGrey",
  '00ced1': "DarkTurquoise",
  '9400d3': "DarkViolet",
  'ff1493': "DeepPink",
  '00bfff': "DeepSkyBlue",
  '696969': "DimGray/DimGrey",
  '1e90ff': "DodgerBlue",
  'b22222': "FireBrick",
  'fffaf0': "FloralWhite",
  '228b22': "ForestGreen",
  'ff00ff': "Fuchsia/Magenta",
  'dcdcdc': "Gainsboro",
  'f8f8ff': "GhostWhite",
  'ffd700': "Gold",
  'daa520': "GoldenRod",
  '808080': "Gray/Grey",
  '008000': "Green",
  'adff2f': "GreenYellow",
  'f0fff0': "HoneyDew",
  'ff69b4': "HotPink",
  'cd5c5c': "IndianRed",
  '4b0082': "Indigo",
  'fffff0': "Ivory",
  'f0e68c': "Khaki",
  'e6e6fa': "Lavender",
  'fff0f5': "LavenderBlush",
  '7cfc00': "LawnGreen",
  'fffacd': "LemonChiffon",
  'add8e6': "LightBlue",
  'f08080': "LightCoral",
  'e0ffff': "LightCyan",
  'fafad2': "LightGoldenRodYellow",
  'd3d3d3': "LightGray/LightGrey",
  '90ee90': "LightGreen",
  'ffb6c1': "LightPink",
  'ffa07a': "LightSalmon",
  '20b2aa': "LightSeaGreen",
  '87cefa': "LightSkyBlue",
  '778899': "LightSlateGray/LightSlateGrey",
  'b0c4de': "LightSteelBlue",
  'ffffe0': "LightYellow",
  '00ff00': "Lime",
  '32cd32': "LimeGreen",
  'faf0e6': "Linen",
  '800000': "Maroon",
  '66cdaa': "MediumAquaMarine",
  '0000cd': "MediumBlue",
  'ba55d3': "MediumOrchid",
  '9370db': "MediumPurple",
  '3cb371': "MediumSeaGreen",
  '7b68ee': "MediumSlateBlue",
  '00fa9a': "MediumSpringGreen",
  '48d1cc': "MediumTurquoise",
  'c71585': "MediumVioletRed",
  '191970': "MidnightBlue",
  'f5fffa': "MintCream",
  'ffe4e1': "MistyRose",
  'ffe4b5': "Moccasin",
  'ffdead': "NavajoWhite",
  '000080': "Navy",
  'fdf5e6': "OldLace",
  '808000': "Olive",
  '6b8e23': "OliveDrab",
  'ffa500': "Orange",
  'ff4500': "OrangeRed",
  'da70d6': "Orchid",
  'eee8aa': "PaleGoldenRod",
  '98fb98': "PaleGreen",
  'afeeee': "PaleTurquoise",
  'db7093': "PaleVioletRed",
  'ffefd5': "PapayaWhip",
  'ffdab9': "PeachPuff",
  'cd853f': "Peru",
  'ffc0cb': "Pink",
  'dda0dd': "Plum",
  'b0e0e6': "PowderBlue",
  '800080': "Purple",
  '663399': "RebeccaPurple",
  'ff0000': "Red",
  'bc8f8f': "RosyBrown",
  '4169e1': "RoyalBlue",
  '8b4513': "SaddleBrown",
  'fa8072': "Salmon",
  'f4a460': "SandyBrown",
  '2e8b57': "SeaGreen",
  'fff5ee': "SeaShell",
  'a0522d': "Sienna",
  'c0c0c0': "Silver",
  '87ceeb': "SkyBlue",
  '6a5acd': "SlateBlue",
  '708090': "SlateGray/SlateGrey",
  'fffafa': "Snow",
  '00ff7f': "SpringGreen",
  '4682b4': "SteelBlue",
  'd2b48c': "Tan",
  '008080': "Teal",
  'd8bfd8': "Thistle",
  'ff6347': "Tomato",
  '40e0d0': "Turquoise",
  'ee82ee': "Violet",
  'f5deb3': "Wheat",
  'ffffff': "White",
  'f5f5f5': "WhiteSmoke",
  'ffff00': "Yellow",
  '9acd32': "YellowGreen",
};


lib/color_parser.dart
library color_parser;

import 'dart:math' as Math;

import 'package:flutter/material.dart';

import 'color_names.dart';

/// class to parse the color and get the color value using different approaches
/// such as RGB, HEX etc.
/// inspired from the JS https://www.w3schools.com/lib/w3color.js
class ColorParser {
  /// color
  Color? _color;

  /// hue
  num? _hue;

  /// saturation
  num? _sat;

  /// lightness
  num? _lightness;

  /// whiteness
  num? _whiteness;

  /// blackness
  num? _blackness;

  /// cyan
  num? _cyan;

  /// magenta
  num? _magenta;

  /// yellow
  num? _yellow;

  /// black
  num? _black;

  /// Ncol
  String? _ncol;

  /// create a object of this class using the RGB values
  ///
  /// r = red value [0-255]
  /// g = green value [0-255]
  /// b = blue value [0-255]
  ///
  /// here the alpha is fully opaque [255]
  ColorParser.rgb(int r, int g, int b) {
    this._color = Color.fromARGB(255, r, g, b);
    _calculateFromRGB();
  }

  /// create a object of this class using the ARGB values
  ///
  /// a = alpha value [0-255], 0 being fully transparent and 255 being fully opaque
  /// r = red value [0-255]
  /// g = green value [0-255]
  /// b = blue value [0-255]
  ColorParser.argb(int a, int r, int g, int b) {
    this._color = Color.fromARGB(a, r, g, b);
    _calculateFromRGB();
  }

  /// create a object of this class using the RGBO values
  ///
  /// r = red value [0-255]
  /// g = green value [0-255]
  /// b = blue value [0-255]
  /// o = opacity as double, with 0.0 being transparent and 1.0 being fully opaque.
  ColorParser.rgbo(int r, int g, int b, double o) {
    this._color = Color.fromRGBO(r, g, b, o);
    _calculateFromRGB();
  }

  /// get the object of this class using material color
  ColorParser.color(Color color) {
    this._color = color;
    _calculateFromRGB();
  }

  /// get the object of this class using int value of color
  ColorParser.value(int value) {
    this._color = Color(value);
    _calculateFromRGB();
  }

  /// get the object of this class using hex color code
  ColorParser.hex(String hexCode) {
    bool startsWithHash = hexCode.startsWith("#");
    this._color = new Color(int.parse(
            hexCode.substring(startsWithHash ? 1 : 0, startsWithHash ? 7 : 6),
            radix: 16) +
        0xFF000000);
    _calculateFromRGB();
  }

  /// calculate values from RGB
  void _calculateFromRGB() {
    List<num> hwb = _calculateHWB(_color!.red, _color!.green, _color!.blue);
    List<num> hsl = _calculateHSL(_color!.red, _color!.green, _color!.blue);
    List<num> cmyk = _calculateCMYK(_color!.red, _color!.green, _color!.blue);

    _hue = num.parse(hsl[0].toStringAsFixed(0));
    _whiteness = num.parse(hwb[1].toStringAsFixed(2));
    _blackness = num.parse(hwb[2].toStringAsFixed(2));
    _sat = num.parse(hsl[1].toStringAsFixed(2));
    _lightness = num.parse(hsl[2].toStringAsFixed(2));
    _cyan = num.parse(cmyk[0].toStringAsFixed(2));
    _magenta = num.parse(cmyk[1].toStringAsFixed(2));
    _yellow = num.parse(cmyk[2].toStringAsFixed(2));
    _black = num.parse(cmyk[3].toStringAsFixed(2));

    _ncol = _hueToNcol(_hue!);
  }

  /// calculate the HWB value from RGB
  List<num> _calculateHWB(int red, int green, int blue) {
    num h, w, bl;

    num r = red / 255;
    num g = green / 255;
    num b = blue / 255;

    num max = Math.max(Math.max(r, g), b);
    num min = Math.min(Math.min(r, g), b);
    num chroma = max - min;
    if (chroma == 0) {
      h = 0;
    } else if (r == max) {
      h = (((g - b) / chroma) % 6) * 360;
    } else if (g == max) {
      h = ((((b - r) / chroma) + 2) % 6) * 360;
    } else {
      h = ((((r - g) / chroma) + 4) % 6) * 360;
    }
    w = min;
    bl = 1 - max;

    return [h, w, bl];
  }

  /// calculate HSL from RGB
  List<num> _calculateHSL(int red, int green, int blue) {
    num? h, s, l;

    List<num> rgb = [];
    rgb.add(red / 255);
    rgb.add(green / 255);
    rgb.add(blue / 255);

    num min = rgb[0];
    num max = rgb[0];
    num maxColor = 0;

    for (int i = 0; i < rgb.length - 1; i++) {
      if (rgb[i + 1] <= min) {
        min = rgb[i + 1];
      }
      if (rgb[i + 1] >= max) {
        max = rgb[i + 1];
        maxColor = i + 1;
      }
    }
    if (maxColor == 0) {
      h = (rgb[1] - rgb[2]) / (max - min);
    }
    if (maxColor == 1) {
      h = 2 + (rgb[2] - rgb[0]) / (max - min);
    }
    if (maxColor == 2) {
      h = 4 + (rgb[0] - rgb[1]) / (max - min);
    }

    if (h!.isNaN) {
      h = 0;
    }
    h = h * 60;
    if (h < 0) {
      h = h + 360;
    }
    l = (min + max) / 2;
    if (min == max) {
      s = 0;
    } else {
      if (l < 0.5) {
        s = (max - min) / (max + min);
      } else {
        s = (max - min) / (2 - max - min);
      }
    }
    return [h, s, l];
  }

  /// calculate CMYK from RGB
  List<num> _calculateCMYK(int red, int green, int blue) {
    num c, m, y, k;
    num r = red / 255;
    num g = green / 255;
    num b = blue / 255;

    num max = Math.max(Math.max(r, g), b);
    k = 1 - max;
    if (k == 1) {
      c = 0;
      m = 0;
      y = 0;
    } else {
      c = (1 - r - k) / (1 - k);
      m = (1 - g - k) / (1 - k);
      y = (1 - b - k) / (1 - k);
    }
    return [c, m, y, k];
  }

  /// calculate Ncol from hue
  String _hueToNcol(num hue) {
    while (hue >= 360) {
      hue = hue - 360;
    }
    if (hue < 60) {
      return "R" + (hue / 0.6).round().toString();
    }
    if (hue < 120) {
      return "Y" + ((hue - 60) / 0.6).round().toString();
    }
    if (hue < 180) {
      return "G" + ((hue - 120) / 0.6).round().toString();
    }
    if (hue < 240) {
      return "C" + ((hue - 180) / 0.6).round().toString();
    }
    if (hue < 300) {
      return "B" + ((hue - 240) / 0.6).round().toString();
    }
    if (hue < 360) {
      return "M" + ((hue - 300) / 0.6).round().toString();
    }

    return "";
  }

  /// get the Color
  Color? getColor() {
    return _color;
  }

  /// get RGB
  List<int> toRGB() {
    return [_color!.red, _color!.green, _color!.blue];
  }

  /// RGB string
  String toRGBString() {
    return "rgb(${_color!.red}, ${_color!.green}, ${_color!.blue})";
  }

  /// get RGB with alpha
  List<int> toRGBA() {
    return [_color!.red, _color!.green, _color!.blue, _color!.alpha];
  }

  /// RGB string with alpha
  String toRGBAString() {
    return "rgba(${_color!.red}, ${_color!.green}, ${_color!.blue}, ${_color!.alpha})";
  }

  /// get HWB
  List<num?> toHWB() {
    return [_hue, _whiteness, _blackness];
  }

  /// HWB string
  String toHWBString() {
    return "hwb($_hue, ${(_whiteness! * 100).round()}%, ${(_blackness! * 100).round()}%)";
  }

  /// HWB string decimal
  String toHWBDecimal() {
    return "hwb($_hue, $_whiteness, $_blackness)";
  }

  /// get HWB with alpha
  List<num?> toHWBA() {
    return [_hue, _whiteness, _blackness, _color!.alpha];
  }

  /// HWB string with alpha
  String toHWBAString() {
    return "hwba($_hue, ${(_whiteness! * 100).round()}%, ${(_blackness! * 100).round()}%, ${_color!.alpha})";
  }

  /// get HSL
  List<num?> toHSL() {
    return [_hue, _sat, _lightness];
  }

  /// HSL string
  String toHSLString() {
    return "hsl($_hue, ${(_sat! * 100).round()}%, ${(_lightness! * 100).round()}%)";
  }

  /// HSL string decimal
  String toHSLDecimal() {
    return "hsl($_hue, $_sat, $_lightness)";
  }

  /// get HSL with alpha
  List<num?> toHSLA() {
    return [
      _hue,
      _sat,
      _lightness,
      _color!.alpha,
    ];
  }

  /// HSL string with alpha
  String toHSLAString() {
    return "hsla($_hue, ${(_sat! * 100).round()}%, ${(_lightness! * 100).round()}%, ${_color!.alpha})";
  }

  /// getCMYK
  List<num?> toCMYK() {
    return [_cyan, _magenta, _yellow, _black];
  }

  /// CMYK string
  String toCMYKString() {
    return "cmyk(${(_cyan! * 100).round()}%, ${(_magenta! * 100).round()}%, ${(_yellow! * 100).round()}%, ${(_black! * 100).round()}%)";
  }

  /// CMYK string decimal
  String toCMYKDecimal() {
    return "cmyk($_cyan, $_magenta, $_yellow, $_black)";
  }

  /// get CMYK with alpha
  List<num?> toCMYKA() {
    return [
      _cyan,
      _magenta,
      _yellow,
      _black,
      _color!.alpha,
    ];
  }

  /// get NCOL
  List<dynamic> toNcol() {
    return [_ncol, _whiteness, _blackness];
  }

  /// NCOL string
  String toNcolString() {
    return "$_ncol, ${(_whiteness! * 100).round()}%, ${(_blackness! * 100).round()}%";
  }

  /// NCOL string decimal
  String toNcolDecimal() {
    return "$_ncol, $_whiteness, $_blackness";
  }

  /// get NCOL with alpha
  List<dynamic> toNcolA() {
    return [_ncol, _whiteness, _blackness, _color!.alpha];
  }

  /// NCOL string with alpha
  String toNcolAString() {
    return "$_ncol, ${(_whiteness! * 100).round()}%, ${(_blackness! * 100).round()}%, ${_color!.alpha}";
  }

  /// get color name
  String? toName() {
    List<String> colorCodes = colorNames.keys.toList();
    for (String code in colorCodes) {
      num r = int.parse(code.substring(0, 2), radix: 16);
      num g = int.parse(code.substring(2, 4), radix: 16);
      num b = int.parse(code.substring(4, 6), radix: 16);

      if (_color!.red == r && _color!.green == g && _color!.blue == b) {
        return colorNames[code];
      }
    }

    return "";
  }

  /// hex string
  String toHex() {
    return "#${_intToHex(_color!.red)}${_intToHex(_color!.green)}${_intToHex(_color!.blue)}";
  }

  /// convert int value to hex
  String _intToHex(int value) {
    String hex = value.toRadixString(16);
    while (hex.length < 2) {
      hex = "0" + hex;
    }
    return hex;
  }
}


example/lib/example.dart
import 'package:color_parser/color_parser.dart';
import 'package:flutter/material.dart';

class Example extends StatefulWidget {
  @override
  _ExampleState createState() => _ExampleState();
}

class _ExampleState extends State<Example> {
  ColorParser? _parser;
  TextEditingController? _controller;
  String _error = "";

  @override
  void initState() {
    // initially use this color
    String code = "#00bfff";
    _parser = ColorParser.hex(code);
    _controller = TextEditingController(text: code);

    super.initState();
  }

  @override
  void dispose() {
    _controller?.dispose();
    super.dispose();
  }

  /// parse the color
  void _parseColor() {
    try {
      setState(() {
        _parser = ColorParser.hex(_controller!.text);
        _error = "";
      });
    } catch (e) {
      setState(() {
        _error = e.toString();
        _parser = null;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black12,
      body: Center(
        child: Container(
          color: Colors.white,
          constraints: BoxConstraints(
            maxWidth: 700,
          ),
          child: Column(
            children: [
              Expanded(
                child: ListView(
                  padding: EdgeInsets.all(16),
                  children: [
                    Text(
                      "Enter Color Hex Code",
                      style: TextStyle(
                        fontSize: 20,
                        color: Colors.black87,
                      ),
                    ),
                    Text(
                      "For this example I have used hex code only, but there are several options to parse the color.\ncheck the documentation at https://github.com/deepak786/color_parser#usage",
                      style: TextStyle(
                        fontSize: 12,
                        color: Colors.black45,
                      ),
                    ),
                    SizedBox(height: 5),
                    TextField(
                      controller: _controller,
                      decoration: InputDecoration(
                        hintText: "hex code e.g (#00bfff)",
                      ),
                    ),
                    if (_error.isNotEmpty)
                      // show the error
                      Text(
                        _error,
                        style: TextStyle(
                          fontSize: 12,
                          color: Colors.red,
                        ),
                      ),
                    SizedBox(height: 5),
                    ElevatedButton(
                      style: ElevatedButton.styleFrom(
                        primary: Theme.of(context).primaryColor,
                      ),
                      onPressed: _parseColor,
                      child: Text(
                        "Parse Color",
                        style: TextStyle(
                          color: Colors.white,
                          fontSize: 15,
                        ),
                      ),
                    ),
                    SizedBox(height: 40),
                    // show the parsed values
                    parsedData(),
                  ],
                ),
              ),
              // flutter love
              Text(
                "Made with ❤️ in Flutter \n",
                style: TextStyle(
                  fontSize: 16,
                  color: Colors.black26,
                ),
                textAlign: TextAlign.center,
              ),
            ],
          ),
        ),
      ),
    );
  }

  /// display the parsed data from parser
  Widget parsedData() {
    if (_parser == null) return Container();

    return Column(
      children: [
        getRow("Name", _parser!.toName() ?? ""),
        divider(),
        Padding(
          padding: const EdgeInsets.only(top: 5, bottom: 5),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.start,
            children: [
              Expanded(
                child: Text(
                  "Color",
                  style: TextStyle(
                    fontSize: 16,
                    fontWeight: FontWeight.bold,
                    color: Colors.black54,
                  ),
                ),
              ),
              Expanded(
                child: Container(
                  height: 100,
                  child: Align(
                    alignment: Alignment.topLeft,
                    child: SizedBox(
                      width: 100,
                      height: 100,
                      child: Container(color: _parser!.getColor()),
                    ),
                  ),
                ),
              ),
            ],
          ),
        ),
        divider(),
        getRow("RGB", _parser!.toRGBString()),
        divider(),
        getRow("Hex", _parser!.toHex()),
        divider(),
        getRow("HSL", _parser!.toHSLString()),
        divider(),
        getRow("HWB", _parser!.toHWBString()),
        divider(),
        getRow("CMYK", _parser!.toCMYKString()),
        divider(),
        getRow("Ncol", _parser!.toNcolString()),
      ],
    );
  }

  /// get the row
  Widget getRow(String title, String value) {
    return Padding(
      padding: const EdgeInsets.only(top: 5, bottom: 5),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.start,
        children: [
          Expanded(
            child: Text(
              title,
              style: TextStyle(
                fontSize: 16,
                fontWeight: FontWeight.bold,
                color: Colors.black54,
              ),
            ),
          ),
          Expanded(
            child: Text(
              value,
              style: TextStyle(
                fontSize: 16,
                color: Colors.black54,
              ),
            ),
          ),
        ],
      ),
    );
  }

  /// divider
  Widget divider() {
    return Container(
      height: 0.5,
      color: Colors.black26,
    );
  }
}
